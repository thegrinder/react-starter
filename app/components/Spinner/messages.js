import { defineMessages } from 'react-intl';

export default defineMessages({
  spinner: {
    id: 'app.components.Spinner',
    defaultMessage: 'loading...',
  },
});
