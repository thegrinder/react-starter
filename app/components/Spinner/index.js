import styled, { keyframes } from 'styled-components';
import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import { getColor, rem } from 'basic-styled-uikit';

import messages from './messages';

const propTypes = {
  sizing: PropTypes.number,
  color: PropTypes.string.isRequired,
};

const defaultProps = {
  sizing: 20,
};

const loading = keyframes`
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
`;

const Spinner = styled.span.attrs(({ intl }) => ({
  'aria-label': intl.formatMessage(messages.spinner),
}))`
  width: ${(props) => rem(props.sizing)(props)};
  height: ${(props) => rem(props.sizing)(props)};
  border-radius: 100%;
  border: 2px solid ${(props) => getColor(props.color)(props)};
  border-bottom-color: transparent;
  animation: ${loading} 0.75s 0s infinite linear;
  vertical-align: middle;
  display: inline-block;
`;

Spinner.propTypes = propTypes;
Spinner.defaultProps = defaultProps;

export default injectIntl(Spinner);
