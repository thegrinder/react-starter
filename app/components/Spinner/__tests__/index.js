import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render } from 'react-testing-library';

import theme from '../../../theme/components';
import Spinner from '..';

const requiredProps = {
  color: 'black',
};

const renderComponent = (props = {}) => render(
  <ThemeProvider theme={theme}>
    <Spinner {...requiredProps} {...props} />
  </ThemeProvider>,
);

describe('<Spinner />', () => {
  it('should render correctly with default props', () => {
    const { container: { firstChild } } = renderComponent();
    expect(firstChild).toBeDefined();
    expect(firstChild).toMatchSnapshot();
  });

  it('should render correctly with custom props', () => {
    const { container: { firstChild } } = renderComponent({ color: 'white', sizing: 40 });
    expect(firstChild).toBeDefined();
  });

  it('should render <span> tag', () => {
    const { container: { firstChild } } = renderComponent();
    expect(firstChild.tagName).toEqual('SPAN');
  });
});
