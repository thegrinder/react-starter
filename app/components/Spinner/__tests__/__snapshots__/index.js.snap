// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spinner /> should render correctly with default props 1`] = `
<span
  class="sc-bdVaJa ioELWI"
  color="black"
/>
`;
